### ISSUE TITLE
nativescript support breaks browser in react-native

### ISSUE THREAD

**Issue by arolson101** [typeorm/typeorm#2531]
**Issue type:**

[ ] question
[X] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[X] `react-native`
[ ] `expo`

**TypeORM version:**

[X] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
I've got a react-native application using typeorm.  I cloned the repository (trying to fix https://github.com/typeorm/typeorm/issues/2502) and the package I build uses nativescript-sqlite though I don't want to use it.

```
Loading dependency graph, done.
error: bundling failed: Error: Unable to resolve module `nativescript-sqlite` from `C:\crap\node_modules\typeorm\build\package\browser\driver\nativescript\NativescriptDriver.js`: Module `nativescript-sqlite` does not exist in the Haste module map
```

**Comment by pleerock** [comment]
thanks for the report we are going to fix it

**Comment by championswimmer** [comment]
requires are not supported in browser runtime https://github.com/typeorm/typeorm/blob/master/src/platform/BrowserPlatformTools.template#L33 
so how exactly do we work around this ?

**Comment by championswimmer** [comment]
ALso @arolson101 did this error occur on `typeorm@latest` ? If I am not wrong `typeorm@0.2.7` is latest on npmjs.com and that version doesn't even have nativescript support on it.

**Comment by pleerock** [comment]
@championswimmer he build repository manually. 

Yeap so I have another report that tells that typeorm in browser has same issue with react native when he simply runs in browser. This issue should be fixed for both react-native and nativescript.

@daniel-lang but I guess we cannot replace to BrowserPlatformTools.load in react native (and nativescript) right? Is react-native uses `typeorm/browser` ? Looks like a serious issue.

**Comment by arolson101** [comment]
@championswimmer I’m using a forked version of the latest to fix a (different) bug.

**Comment by arolson101** [comment]
@pleerock yes, react-native uses typeorm/browser

**Comment by championswimmer** [comment]
We need to whitelist the drivers in Webpack so that webpack doesn't try to parse the require statements for optional modules.

**Comment by championswimmer** [comment]
Possible fix here https://github.com/typeorm/typeorm/pull/2553

**Comment by championswimmer** [comment]
@arolson101 can you try with this build and report

**Comment by arolson101** [comment]
checking now.  FYI, I still see these errors:

```
build\browser\src\find-options\FindOptionsUtils.ts(92,58): error TS2345: Argument of type 'keyof T' is not assignable to parameter of type 'string'.
  Type 'string | number | symbol' is not assignable to type 'string'.
    Type 'number' is not assignable to type 'string'.
TypeScript: 1 semantic error
TypeScript: emit succeeded (with errors)
src\find-options\FindOptionsUtils.ts(92,58): error TS2345: Argument of type 'keyof T' is not assignable to parameter of type 'string'.
  Type 'string | number | symbol' is not assignable to type 'string'.
    Type 'number' is not assignable to type 'string'.
```

**Comment by arolson101** [comment]
@championswimmer 
same error:
```
error: bundling failed: Error: Unable to resolve module `nativescript-sqlite` from `C:\crap\node_modules\typeorm\build\package\browser\driver\nativescript\NativescriptDriver.js`: Module `nativescript-sqlite` does not exist in the Haste module map
```

FWIW https://github.com/typeorm/typeorm/issues/2479 would be a good solution to this

**Comment by championswimmer** [comment]
How about adding `nativescript-sqlite` to webpack externals? That fixes this right ?

**Comment by championswimmer** [comment]
@arolson101 @cuibonobo 
Can you check this again ? 
I just tried it with webpack, and https://github.com/typeorm/typeorm/pull/2553 works perfectly fine. Generates error on runtime only, if the specified module isn't present.

So you do not need nativescript-sqlite to run react native (and vice versa)

Webpack can only parse `require()` if it has a constant string argument.

**Comment by championswimmer** [comment]
How about, for the browser based drivers, we make the user require the driver in his code and just pass the required object into Typeorm createConnection() options ?? 

```typescript
createConnection({
  type: 'nativescript'
  driver: require('nativescript-sqlite')
})
```

or 

```typescript
import driver from 'nativescript-sqlite'
createConnection({
  type: 'nativescript'
  driver: driver
})
```

**Comment by championswimmer** [comment]
@pleerock  @daniel-lang @arolson101 
Let me know what do you think about that ?

**Comment by arolson101** [comment]
@championswimmer having the db library require be in client code would be fantastic; that looks like a good way to do it.

I haven’t been able to check yet, but note that react-native doesn’t use Webpack, so those tricks are unlikely to work.

**Comment by championswimmer** [comment]
@daniel-lang @pleerock 
did you think about this 
https://github.com/typeorm/typeorm/issues/2531#issuecomment-411608720

**Comment by championswimmer** [comment]
@daniel-lang @pleerock 
any updates ?

**Comment by championswimmer** [comment]
following types - 

`react-native`
`nativescript`
`cordova`
`expo`

**Comment by Kononnable** [comment]
@cuibonobo 
If nothing unexpected happened during this two weeks(and your branch solves this issue) I think you should make a PR. It would be easier to convince more people to test that solution and release `0.2.8`. If it breaks code in some specific way PR can always be fixed or rejected.

However won't it break existing code? I see some require statements going away. If that's true, then it is extension for next branch and on `0.2.x` we still haven't got that problem figured out.

**Comment by arolson101** [comment]
This would be breaking, but doesn’t have to be- just create a new library entry point ‘typeorm/base’ or something that exports this new createConnection function, then have ‘typeorm/index’ export a wrapper function that does the requires for backwards compatibility.

**Comment by cuibonobo** [comment]
Yeah, the solution that I came up with would break existing code, but I'm on the fence because my code is broken *now* because of the changes introduced in #2407. My build complains about `nativescript-sqlite` not being available and even if I `npm install` it, my code still won't run.

I think the fact that TypeScript would now complain about a missing `driver` option for browser implementations would be enough of a clue for developers to look at the documentation and see what has changed.

**Comment by championswimmer** [comment]
ideally we should externalise all the browser drivers and make users provide it to us,
**BUT** we should unbreak the breaking changes ONLY nativescript introduced, without having to make semver-major release

https://github.com/typeorm/typeorm/pull/2854 

This PR should enable use to make a v0.2.8 release

**Comment by Kononnable** [comment]
@cuibonobo Could you check if your code works with fix in #2854 ? It doesn't look breaking so it could be merged with 0.2.x releases.

**Comment by championswimmer** [comment]
I think we can have a 0.2.8 release finally now.
Thanks @pleerock for merging
